File: python/intervals/minimum-interval-to-include-each-query.py
import heapq


class Solution:
    def minInterval(self, intervals: List[List[int]], queries: List[int]) -> List[int]:
        answer = []
        lengths = {}
        heap = []

        intervals.sort()
        sorted_queries = queries[:]
        sorted_queries.sort()

        count = 0
        for q in sorted_queries:
            for i in range(count, len(intervals)):
                if intervals[i][0] <= q:
                    heapq.heappush(heap, [intervals[i][1] - intervals[i][0] + 1, intervals[i][0], intervals[i][1]])
                else:
                    count = i
                    break

            if len(heap) == 0:
                lengths[q] = -1
                continue

            current = heapq.heappop(heap)
            while heap and current[2] < q:
                current = heapq.heappop(heap)
            heapq.heappush(heap, current)

            if q <= current[2] and q >= current[1]:
                lengths[q] = current[0]
            else:
                lengths[q] = -1

        for q in queries:
            answer.append(lengths[q])

        return answer

File: c++/intervals/minimum-interval-to-include-each-query.cc
class Solution {
public:
    vector<int> minInterval(vector<vector<int>>& intervals, vector<int>& queries) {
        std::sort(intervals.begin(), intervals.end());

        std::unordered_map<int, int> indexes;
        for (int i = 0; i < queries.size(); ++i) {
            indexes[i] = queries[i];
        }

        std::sort(queries.begin(), queries.end());
        int index = 0;
        std::priority_queue<std::vector<int>> pq;
        std::unordered_map<int, int> results;
        for (int query : queries) {
            results[query] = -1;
            while (index < intervals.size() && intervals[index][0] <= query) {
                pq.push(std::vector<int>{-(intervals[index][1] - intervals[index][0] + 1), intervals[index][0], intervals[index][1]});
                index++;
            }

            while (pq.size() > 0) {
                std::vector<int> current = pq.top();
                if (current[2] < query) {
                    pq.pop();
                } else {
                    results[query] = -current[0];
                    break;
                }
            }
        }

        std::vector<int> answer;
        for (int i = 0; i < queries.size(); ++i) {
            answer.push_back(results[indexes[i]]);
        }
        
        return answer;
    }
};
